<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Philipp SL Schäfer" />


<title>Covid-19 BALF scRNA-seq Quality Control</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Bioinfo Playground</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Content
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="iterative_lsi.html">Iterative LSI</a>
    </li>
    <li>
      <a href="integration_cca_2.html">Integration via CCA</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Covid-19 BALF scRNA-seq Quality Control</h1>
<h4 class="author">Philipp SL Schäfer</h4>
<h4 class="date">2022-04-11-14-09</h4>

</div>


<div id="status" class="section level1">
<h1>Status</h1>
<p>Work in progress</p>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<pre class="r"><code>knitr::opts_chunk$set(echo = TRUE)
library(reticulate)
Sys.setenv(RETICULATE_PYTHON =
             &quot;/home/pschaefer/miniconda3/envs/r-reticulate/bin/python&quot;)
timestamp &lt;- format(Sys.time(), &quot;%Y-%m-%d-%H-%M&quot;)
s &lt;- 4242
run_again &lt;- TRUE</code></pre>
<pre class="r"><code>if (run_again) {
  print(&quot;run_again set to TRUE, thus everything will be recomputed&quot;)
} else {
  print(&quot;run_again set to FALSE, thus old results will be kept&quot;)
}</code></pre>
<pre><code>## [1] &quot;run_again set to TRUE, thus everything will be recomputed&quot;</code></pre>
<p>Loading Packages</p>
<pre class="r"><code>suppressPackageStartupMessages({
  library(tidyverse)
  library(magrittr)
  library(sparseMatrixStats)
  library(scDblFinder)
  library(scuttle)
  library(scran)
  library(scater)
  library(ggside)
})</code></pre>
</div>
<div id="import-data" class="section level1">
<h1>Import Data</h1>
<pre class="r"><code>if (Sys.info()[&quot;nodename&quot;] == &quot;CrunchyPeanut&quot;) {
  input_dir &lt;- &quot;/mnt/sda/data/cov19_raw/&quot;
  output_dir &lt;- &quot;/home/philipp/Bioinfo_Playground/cov19_balf_gex_qc_output/&quot;
  if (!dir.exists(output_dir)) dir.create(output_dir)
} else {
  warning(&quot;No appropriate nodename, adjust paths manually&quot;)
  stop()
}</code></pre>
<pre class="r"><code>list.files(input_dir)</code></pre>
<pre><code>##  [1] &quot;GSM4339769_C141_filtered_feature_bc_matrix.h5&quot;
##  [2] &quot;GSM4339770_C142_filtered_feature_bc_matrix.h5&quot;
##  [3] &quot;GSM4339771_C143_filtered_feature_bc_matrix.h5&quot;
##  [4] &quot;GSM4339772_C144_filtered_feature_bc_matrix.h5&quot;
##  [5] &quot;GSM4339773_C145_filtered_feature_bc_matrix.h5&quot;
##  [6] &quot;GSM4339774_C146_filtered_feature_bc_matrix.h5&quot;
##  [7] &quot;GSM4475048_C51_filtered_feature_bc_matrix.h5&quot; 
##  [8] &quot;GSM4475049_C52_filtered_feature_bc_matrix.h5&quot; 
##  [9] &quot;GSM4475050_C100_filtered_feature_bc_matrix.h5&quot;
## [10] &quot;GSM4475051_C148_filtered_feature_bc_matrix.h5&quot;
## [11] &quot;GSM4475052_C149_filtered_feature_bc_matrix.h5&quot;
## [12] &quot;GSM4475053_C152_filtered_feature_bc_matrix.h5&quot;
## [13] &quot;hc4&quot;</code></pre>
<p>Reading in the meta data from the authors.</p>
<pre class="r"><code>samples_meta &lt;- read_tsv( url( &quot;https://raw.githubusercontent.com/zhangzlab/covid_balf/master/meta.txt&quot; ) )</code></pre>
<pre><code>## Rows: 13 Columns: 10
## ── Column specification ─────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (5): sample, sample_new, sample_new_old, group, disease
## dbl (5): nCoV_mean, nFeature_RNA_low, nFeature_RNA_high, nCount_RNA_t...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>samples_meta</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["sample"],"name":[1],"type":["chr"],"align":["left"]},{"label":["sample_new"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sample_new_old"],"name":[3],"type":["chr"],"align":["left"]},{"label":["group"],"name":[4],"type":["chr"],"align":["left"]},{"label":["disease"],"name":[5],"type":["chr"],"align":["left"]},{"label":["nCoV_mean"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["nFeature_RNA_low"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["nFeature_RNA_high"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["nCount_RNA_threshold"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["percent.mito"],"name":[10],"type":["dbl"],"align":["right"]}],"data":[{"1":"C51","2":"HC1","3":"HC1","4":"HC","5":"N","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C52","2":"HC2","3":"HC2","4":"HC","5":"N","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C100","2":"HC3","3":"HC3","4":"HC","5":"N","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"GSM3660650","2":"HC4","3":"HC4","4":"HC","5":"N","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C141","2":"M1","3":"O1","4":"M","5":"Y","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C142","2":"M2","3":"O2","4":"M","5":"Y","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C144","2":"M3","3":"O3","4":"M","5":"Y","6":"0.000000000","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C145","2":"S1","3":"S1","4":"S","5":"Y","6":"0.085901131","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C143","2":"S2","3":"C1","4":"S","5":"Y","6":"0.007000048","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C146","2":"S3","3":"C2","4":"S","5":"Y","6":"0.749939188","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C148","2":"S4","3":"C3","4":"S","5":"Y","6":"0.002551020","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C149","2":"S5","3":"C4","4":"S","5":"Y","6":"0.408822508","7":"200","8":"6000","9":"1000","10":"10"},{"1":"C152","2":"S6","3":"C5","4":"S","5":"Y","6":"0.096482152","7":"200","8":"6000","9":"1000","10":"10"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Reading in the matrices.</p>
<pre class="r"><code>rna_list &lt;- purrr::map2(samples_meta$sample, samples_meta$sample_new, function(sample, name) {
  if (!str_starts(sample, &quot;GSM&quot;)) {
    Seurat::Read10X_h5(list.files(input_dir, full.names = TRUE)[
      str_detect(list.files(input_dir), sample)
      ])
  } else {
    Seurat::Read10X(list.files(input_dir, full.names = TRUE)[
      str_detect(list.files(input_dir), tolower(name))
    ])
  }
}) %&gt;%
  purrr::set_names(samples_meta$sample_new)</code></pre>
<pre><code>## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you

## Warning in sparseMatrix(i = indices[] + 1, p = indptr[], x = as.numeric(x
## = counts[]), : &#39;giveCsparse&#39; has been deprecated; setting &#39;repr = &quot;T&quot;&#39;
## for you</code></pre>
<pre class="r"><code>str(rna_list)</code></pre>
<pre><code>## List of 13
##  $ HC1:Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:13825036] 27 66 72 78 93 102 105 122 154 178 ...
##   .. ..@ p       : int [1:11116] 0 1635 3021 4750 6389 8111 8496 9808 10143 11450 ...
##   .. ..@ Dim     : int [1:2] 33538 11115
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33538] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:11115] &quot;AAACCTGAGACACTAA-1&quot; &quot;AAACCTGAGGAGTACC-1&quot; &quot;AAACCTGAGGATATAC-1&quot; &quot;AAACCTGAGGTCATCT-1&quot; ...
##   .. ..@ x       : num [1:13825036] 5 1 1 2 1 1 1 1 9 2 ...
##   .. ..@ factors : list()
##  $ HC2:Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:9313106] 27 76 152 154 161 185 224 326 414 465 ...
##   .. ..@ p       : int [1:10367] 0 786 1076 1606 2634 3562 4575 5453 5878 6930 ...
##   .. ..@ Dim     : int [1:2] 33538 10366
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33538] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:10366] &quot;AAACCTGAGATATACG-1&quot; &quot;AAACCTGAGATCCGAG-1&quot; &quot;AAACCTGAGCACAGGT-1&quot; &quot;AAACCTGAGCGTGAAC-1&quot; ...
##   .. ..@ x       : num [1:9313106] 1 1 1 5 1 1 1 1 1 1 ...
##   .. ..@ factors : list()
##  $ HC3:Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:17357707] 643 1038 1810 1871 1931 1933 1960 2398 3363 3465 ...
##   .. ..@ p       : int [1:8973] 0 83 219 1596 1668 2504 3609 7128 7245 11775 ...
##   .. ..@ Dim     : int [1:2] 33538 8972
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33538] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:8972] &quot;AAACCCAAGGTAGACC-1&quot; &quot;AAACCCACACTGTCCT-1&quot; &quot;AAACCCACAGCTACAT-1&quot; &quot;AAACCCAGTTCGGCCA-1&quot; ...
##   .. ..@ x       : num [1:17357707] 1 1 1 1 1 1 2 1 1 1 ...
##   .. ..@ factors : list()
##  $ HC4:Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:13467593] 19271 28197 32740 7482 26573 33501 29390 31101 33501 5733 ...
##   .. ..@ p       : int [1:737281] 0 3 6 6 9 11 12 14 15 15 ...
##   .. ..@ Dim     : int [1:2] 33538 737280
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33538] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:737280] &quot;AAACCTGAGAAACCAT-1&quot; &quot;AAACCTGAGAAACCGC-1&quot; &quot;AAACCTGAGAAACCTA-1&quot; &quot;AAACCTGAGAAACGAG-1&quot; ...
##   .. ..@ x       : num [1:13467593] 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..@ factors : list()
##  $ M1 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:13620931] 27 51 53 55 121 152 154 201 259 271 ...
##   .. ..@ p       : int [1:6250] 0 1596 6872 6935 8724 8786 10790 10843 10898 12349 ...
##   .. ..@ Dim     : int [1:2] 33539 6249
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:6249] &quot;AAACCTGAGATGTCGG-1&quot; &quot;AAACCTGAGGCTCATT-1&quot; &quot;AAACCTGAGTTGTCGT-1&quot; &quot;AAACCTGCAATCCGAT-1&quot; ...
##   .. ..@ x       : num [1:13620931] 8 1 1 1 1 2 9 1 1 1 ...
##   .. ..@ factors : list()
##  $ M2 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:13152036] 471 688 736 1617 1674 2192 2353 5645 5704 6202 ...
##   .. ..@ p       : int [1:10270] 0 89 1527 1584 6793 8482 12855 12912 17866 17925 ...
##   .. ..@ Dim     : int [1:2] 33539 10269
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:10269] &quot;AAACCTGAGATCGGGT-1&quot; &quot;AAACCTGAGTGAACGC-1&quot; &quot;AAACCTGCAGCTGCAC-1&quot; &quot;AAACCTGGTCAAAGCG-1&quot; ...
##   .. ..@ x       : num [1:13152036] 1 1 1 1 1 1 1 1 1 9 ...
##   .. ..@ factors : list()
##  $ M3 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:1397983] 1498 2339 2405 2592 2895 4431 5000 11451 14397 14690 ...
##   .. ..@ p       : int [1:3717] 0 40 85 124 450 510 593 679 1078 1131 ...
##   .. ..@ Dim     : int [1:2] 33539 3716
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:3716] &quot;AAACCTGAGCTAGTGG-1&quot; &quot;AAACCTGCAGCCTTTC-1&quot; &quot;AAACCTGCAGCTGTGC-1&quot; &quot;AAACCTGGTCTAGTGT-1&quot; ...
##   .. ..@ x       : num [1:1397983] 1 1 1 1 1 2 1 2 2 2 ...
##   .. ..@ factors : list()
##  $ S1 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:19417652] 27 29 77 171 229 251 326 404 473 493 ...
##   .. ..@ p       : int [1:18045] 0 691 1108 1852 3855 5557 6090 6361 7615 8335 ...
##   .. ..@ Dim     : int [1:2] 33539 18044
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:18044] &quot;AAACCTGAGCCCGAAA-1&quot; &quot;AAACCTGAGCCTATGT-1&quot; &quot;AAACCTGAGCTACCGC-1&quot; &quot;AAACCTGAGGAGTAGA-1&quot; ...
##   .. ..@ x       : num [1:19417652] 7 1 1 1 1 1 1 1 1 3 ...
##   .. ..@ factors : list()
##  $ S2 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:35103515] 26 27 66 93 154 178 229 240 261 382 ...
##   .. ..@ p       : int [1:20858] 0 999 2394 3641 4487 7799 9367 10814 13179 13495 ...
##   .. ..@ Dim     : int [1:2] 33539 20857
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:20857] &quot;AAACCTGAGACTTTCG-1&quot; &quot;AAACCTGAGCATGGCA-1&quot; &quot;AAACCTGAGCCATCGC-1&quot; &quot;AAACCTGAGCGTGAAC-1&quot; ...
##   .. ..@ x       : num [1:35103515] 1 12 1 1 1 2 1 1 2 1 ...
##   .. ..@ factors : list()
##  $ S3 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:2812364] 26 27 100 229 451 595 602 736 762 783 ...
##   .. ..@ p       : int [1:4112] 0 296 1858 2184 2500 2721 6477 8770 8827 9341 ...
##   .. ..@ Dim     : int [1:2] 33539 4111
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:4111] &quot;AAACCTGCACATAACC-1&quot; &quot;AAACCTGCACGACTCG-1&quot; &quot;AAACCTGCAGGCTGAA-1&quot; &quot;AAACCTGGTGCACCAC-1&quot; ...
##   .. ..@ x       : num [1:2812364] 1 30 1 1 1 1 1 1 2 1 ...
##   .. ..@ factors : list()
##  $ S4 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:6186401] 493 740 904 918 936 1512 1617 1802 1931 1933 ...
##   .. ..@ p       : int [1:3921] 0 107 6378 6984 7017 9052 10712 15677 17745 19035 ...
##   .. ..@ Dim     : int [1:2] 33539 3920
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:3920] &quot;AAACCTGAGAGAACAG-1&quot; &quot;AAACCTGAGCTGCCCA-1&quot; &quot;AAACCTGAGGAATCGC-1&quot; &quot;AAACCTGGTGGTACAG-1&quot; ...
##   .. ..@ x       : num [1:6186401] 1 1 1 1 1 1 1 1 2 1 ...
##   .. ..@ factors : list()
##  $ S5 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:4251674] 26 27 78 528 531 597 600 607 643 762 ...
##   .. ..@ p       : int [1:2880] 0 297 335 1828 3010 4676 5949 8406 10214 11217 ...
##   .. ..@ Dim     : int [1:2] 33539 2879
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:2879] &quot;AAACCTGAGCGAAGGG-1&quot; &quot;AAACCTGCACCTCGTT-1&quot; &quot;AAACCTGCAGTGGGAT-1&quot; &quot;AAACCTGCATTCGACA-1&quot; ...
##   .. ..@ x       : num [1:4251674] 3 32 1 1 1 2 1 1 2 1 ...
##   .. ..@ factors : list()
##  $ S6 :Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
##   .. ..@ i       : int [1:7923709] 15 27 154 178 251 271 414 449 493 504 ...
##   .. ..@ p       : int [1:7733] 0 630 1045 2140 2350 7868 7915 9021 11075 11511 ...
##   .. ..@ Dim     : int [1:2] 33539 7732
##   .. ..@ Dimnames:List of 2
##   .. .. ..$ : chr [1:33539] &quot;MIR1302-2HG&quot; &quot;FAM138A&quot; &quot;OR4F5&quot; &quot;AL627309.1&quot; ...
##   .. .. ..$ : chr [1:7732] &quot;AAACCTGAGCCACTAT-1&quot; &quot;AAACCTGAGTAAGTAC-1&quot; &quot;AAACCTGAGTGGGCTA-1&quot; &quot;AAACCTGAGTGTTTGC-1&quot; ...
##   .. ..@ x       : num [1:7923709] 1 3 3 1 1 1 1 1 5 1 ...
##   .. ..@ factors : list()</code></pre>
</div>
<div id="subsetting-genes" class="section level1">
<h1>Subsetting Genes</h1>
<p>We need to remove the cov19 row before we can proceed.</p>
<pre class="r"><code>rownames(rna_list$M1)[grep(&quot;cov&quot;, rownames(rna_list$M1), ignore.case=T)]</code></pre>
<pre><code>## [1] &quot;nCoV&quot;</code></pre>
<pre class="r"><code>rna_list &lt;- purrr::map(rna_list, ~ .x[rownames(.x) != &quot;nCoV&quot;, ])
# check that the rownames are consistent
stopifnot(
  all(
    purrr::map_dfc(rna_list, ~rownames(.x)) %&gt;% apply(1, function(x) all(x==x[1]))
    )
)</code></pre>
<p>First we will remove all genes that are not expressed in at least 3 cell in all sample (speeding up the downstream analysis).</p>
<p>Removing genes which are not expressed in at least 3 cells in all samples.</p>
<pre class="r"><code>idx &lt;- purrr::map_dfc(rna_list, ~ Matrix::rowSums(.x != 0) &gt;= 3) %&gt;%
  apply(1, all)
good_genes &lt;- rownames(rna_list[[1]])[idx]
print(paste0(&quot;Retained genes: &quot;, length(good_genes), 
             &quot; from &quot;, nrow(rna_list[[1]])))</code></pre>
<pre><code>## [1] &quot;Retained genes: 13542 from 33538&quot;</code></pre>
<pre class="r"><code>rna_list &lt;- purrr::map(rna_list, function(smp) {
    smp[good_genes, ]
  })</code></pre>
</div>
<div id="empty-droplets" class="section level1">
<h1>Empty Droplets</h1>
<p>Before subsetting we will create a tibble containing the metadata which we will update throughout the analysis.</p>
<pre class="r"><code>coldata &lt;- purrr::imap_dfr(rna_list, ~ tibble::tibble(
  sample = .y, barcode = colnames(.x))) %&gt;%
  mutate(joint_barcode = paste0(sample, &quot;_&quot;, barcode))
coldata %&gt;% glimpse()</code></pre>
<pre><code>## Rows: 845,510
## Columns: 3
## $ sample        &lt;chr&gt; &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, …
## $ barcode       &lt;chr&gt; &quot;AAACCTGAGACACTAA-1&quot;, &quot;AAACCTGAGGAGTACC-1&quot;, &quot;AAAC…
## $ joint_barcode &lt;chr&gt; &quot;HC1_AAACCTGAGACACTAA-1&quot;, &quot;HC1_AAACCTGAGGAGTACC-1…</code></pre>
<p>Plotting log10 of the cumulative UMI counts vs the rank of the cell or rather barcode (“kneeplot”). We need to find thresholds for each sample.</p>
<pre class="r"><code>purrr::imap_dfr(rna_list, ~ tibble(sample = .y, 
                                  umi_per_cell = Matrix::colSums(.x))) %&gt;%
  arrange(sample, desc(umi_per_cell)) %&gt;%
  group_by(sample) %&gt;%
  mutate(idx = seq_along(sample)) %&gt;%
  mutate(cum_umi_per_cell = cumsum(umi_per_cell)) %&gt;%
  ggplot() +
  geom_line(aes(x=idx, y=cum_umi_per_cell)) +
  facet_wrap(~sample, scales = &quot;fixed&quot;) +
  scale_y_log10() + scale_x_log10() +
  labs(x = &quot;Index&quot;, y = &quot;Cumulative UMI Count&quot;)</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-11-1.png" width="1536" /></p>
<p>And we will be using the following thresholds (determined by visual inspection). Especially for “sm027_Tel_5c_20211123” there is a very obvious knee, for the other samples the thresholds are a little bit arbitrary (like everything in bioinformatics).</p>
<pre class="r"><code>thresholds &lt;- c(&quot;HC1&quot; = 10000, 
                &quot;HC2&quot; = 10000,
                &quot;HC3&quot; = 9000,
                &quot;HC4&quot; = 1800,
                &quot;M1&quot; = 3000,
                &quot;M2&quot; = 1800,
                &quot;M3&quot; = 3000,
                &quot;S1&quot; = 10000,
                &quot;S2&quot; = 10000,
                &quot;S3&quot; = 3000,
                &quot;S4&quot; = 1800,
                &quot;S5&quot; = 2200,
                &quot;S6&quot; = 2500
)
stopifnot(all(names(rna_list) == names(thresholds)))</code></pre>
<p>We clearly see that we got the filtered matrices already (meaning filtered by cellranger), except for the additional sample HC4.</p>
<pre class="r"><code>p.list &lt;- purrr::map(unique(coldata$sample), function(smp) {
  tibble::tibble(
    sample = smp,
    umi_per_cell = sparseMatrixStats::colSums2(rna_list[[smp]])) %&gt;%
    arrange(sample, desc(umi_per_cell)) %&gt;%
    group_by(sample) %&gt;%
    mutate(idx = seq_along(sample)) %&gt;%
    mutate(cum_umi_per_cell = cumsum(umi_per_cell)) %&gt;%
    ggplot() +
    geom_line(aes(x=idx, y=cum_umi_per_cell)) +
    geom_vline(xintercept = thresholds[smp], color = &quot;red&quot;) +
    scale_y_log10() + scale_x_log10() +
    labs(x = &quot;Index&quot;, y = &quot;Cumulative UMI Count&quot;, title = smp) +
    theme(plot.title = element_text(hjust = 0.5))
})
do.call(gridExtra::grid.arrange, p.list)</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-13-1.png" width="1536" /></p>
<p>Lastly we subset the count matrices based on these results.</p>
<pre class="r"><code>cell_qc &lt;- purrr::map_dfr(unique(coldata$sample), function(smp) {
  tibble(sample = smp, 
         sum_umis = Matrix::colSums(rna_list[[smp]]),
         detected_genes = Matrix::colSums(rna_list[[smp]] != 0),
         barcode = colnames(rna_list[[smp]]), 
         threshold = thresholds[smp]) }) %&gt;%
  arrange(sample, desc(sum_umis)) %&gt;%
  group_by(sample) %&gt;%
  mutate(rank = seq_along(sample)) %&gt;%
  mutate(cum_umi_per_cell = cumsum(sum_umis)) %&gt;%
  mutate(non_empty = ifelse(rank &lt;= threshold, TRUE, FALSE))

coldata &lt;- coldata %&gt;%
  dplyr::left_join(cell_qc, by=c(&quot;sample&quot;, &quot;barcode&quot;)) %&gt;%
  dplyr::select(sample, barcode, joint_barcode, sum_umis, detected_genes,
         non_empty) %&gt;%
  dplyr::filter(non_empty) %&gt;%
  dplyr::select(!non_empty)

rna_list &lt;- purrr::map(names(rna_list), function(smp) {
  mtx &lt;- rna_list[[smp]]
  valid_barcodes &lt;- coldata %&gt;% 
    dplyr::filter(sample == smp) %&gt;% 
    dplyr::pull(barcode)
  mtx[,valid_barcodes]
}) %&gt;% set_names(names(rna_list))</code></pre>
<p>Check again. And I think it looks ok now.</p>
<pre class="r"><code>purrr::imap_dfr(rna_list, ~ tibble(sample = .y, 
                                  umi_per_cell = Matrix::colSums(.x))) %&gt;%
  arrange(sample, desc(umi_per_cell)) %&gt;%
  group_by(sample) %&gt;%
  mutate(idx = seq_along(sample)) %&gt;%
  mutate(cum_umi_per_cell = cumsum(umi_per_cell)) %&gt;%
  ggplot() +
  geom_line(aes(x=idx, y=cum_umi_per_cell)) +
  facet_wrap(~sample, scales = &quot;fixed&quot;) +
  scale_y_log10() + scale_x_log10() +
  labs(x = &quot;Index&quot;, y = &quot;Cumulative UMI Count&quot;)</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-15-1.png" width="1536" /></p>
</div>
<div id="rna-doublet-scores---scdblfinder" class="section level1">
<h1>RNA Doublet Scores - <code>scDblFinder</code></h1>
<p>We will use <code>scDblFinder</code> as this package it implemented in R and seems to be better than <code>scrublet</code> (according to their vignette at least, but I guess their comparison might be biased!). Usually these tools simulate doublets and check how similar your actual cells are to these simulated doublets.</p>
<p>See the description/vignette of the tool here:</p>
<ol style="list-style-type: decimal">
<li><p><a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scDblFinder/inst/doc/computeDoubletDensity.html" class="uri">http://www.bioconductor.org/packages/release/bioc/vignettes/scDblFinder/inst/doc/computeDoubletDensity.html</a></p></li>
<li><p><a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scDblFinder/inst/doc/scDblFinder.html" class="uri">http://www.bioconductor.org/packages/release/bioc/vignettes/scDblFinder/inst/doc/scDblFinder.html</a></p></li>
</ol>
<p>To use the tools we will temporarily create a single cell experiment object.</p>
<pre class="r"><code>dbr &lt;- 0.08
print(paste0(&quot;Using an expected doublet rate of: &quot;, dbr))</code></pre>
<pre><code>## [1] &quot;Using an expected doublet rate of: 0.08&quot;</code></pre>
<pre class="r"><code>doublet_scores_scdbl &lt;- purrr::imap(rna_list, function(smp, name) {
  
  print(name)
  sce &lt;- SingleCellExperiment(list(counts = smp))
  sce &lt;- logNormCounts(sce)
  dec &lt;- modelGeneVar(sce)
  hvgs &lt;- getTopHVGs(dec, n=2000)
  set.seed(s)
  sce &lt;- runPCA(sce, ncomponents=30, subset_row=hvgs)
  sce &lt;- runUMAP(sce, dimred=&quot;PCA&quot;)
  #scores &lt;- computeDoubletDensity(sce, subset.row=hvgs)
  sce &lt;- scDblFinder(sce, dbr=dbr, iter=1)
  scores &lt;- sce$scDblFinder.score
  classification &lt;- sce$scDblFinder.class
  names(scores) &lt;- colnames(smp)
  names(classification) &lt;- colnames(smp)

  # prepare plots
  p1 &lt;- tibble::tibble(cell = names(scores), score = scores, 
                 UMAP1 = reducedDim(sce, &quot;UMAP&quot;)[,1],
                 UMAP2 = reducedDim(sce, &quot;UMAP&quot;)[,2]) %&gt;%
    ggplot() +
    geom_point(aes(x=UMAP1, y=UMAP2, color=score), size=0.2) +
    scale_color_viridis_c()
  p2 &lt;- ggplot() +
    geom_histogram(aes(scores), bins = 100, fill=&quot;grey&quot;, color=&quot;black&quot;) +
    labs(y = &quot;Count&quot;, x = &quot;Doublet Score&quot;)
  
  # use cowplot for plotting the grid
  plot_row &lt;- cowplot::plot_grid(p1, p2)
  title &lt;- cowplot::ggdraw() + 
    cowplot::draw_label(label = paste(name), fontface = &quot;bold&quot;, x = 0, hjust = -2)
  print(cowplot::plot_grid(title, plot_row, ncol = 1, rel_heights = c(0.1, 1)))
  list(score = scores, classifcation = classification)
})</code></pre>
<pre><code>## [1] &quot;HC1&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-1.png" width="960" /></p>
<pre><code>## [1] &quot;HC2&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-2.png" width="960" /></p>
<pre><code>## [1] &quot;HC3&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-3.png" width="960" /></p>
<pre><code>## [1] &quot;HC4&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-4.png" width="960" /></p>
<pre><code>## [1] &quot;M1&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-5.png" width="960" /></p>
<pre><code>## [1] &quot;M2&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-6.png" width="960" /></p>
<pre><code>## [1] &quot;M3&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-7.png" width="960" /></p>
<pre><code>## [1] &quot;S1&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-8.png" width="960" /></p>
<pre><code>## [1] &quot;S2&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-9.png" width="960" /></p>
<pre><code>## [1] &quot;S3&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-10.png" width="960" /></p>
<pre><code>## [1] &quot;S4&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-11.png" width="960" /></p>
<pre><code>## [1] &quot;S5&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-12.png" width="960" /></p>
<pre><code>## [1] &quot;S6&quot;</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-16-13.png" width="960" /></p>
<p>Check how many predicted doublets per sample using our threshold set above. Let’s keep the doublet scores and we will not remove/mask any cells, but rather check later on whether doublets are particularly enriched in any cluster we get.</p>
<pre class="r"><code>doublet_scores_scdbl %&gt;%
  purrr::imap_dfr(function(smp, name) {
    tibble::tibble(sample = name, doublet_score = smp[[&quot;score&quot;]],
                   predicted_doublet = smp[[&quot;classifcation&quot;]])
  }) %&gt;%
  dplyr::count(sample, predicted_doublet) %&gt;%
  pivot_wider(names_from=predicted_doublet, values_from=n)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["sample"],"name":[1],"type":["chr"],"align":["left"]},{"label":["singlet"],"name":[2],"type":["int"],"align":["right"]},{"label":["doublet"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"HC1","2":"8871","3":"1129"},{"1":"HC2","2":"9111","3":"889"},{"1":"HC3","2":"8057","3":"915"},{"1":"HC4","2":"1658","3":"142"},{"1":"M1","2":"2845","3":"155"},{"1":"M2","2":"1653","3":"147"},{"1":"M3","2":"2773","3":"227"},{"1":"S1","2":"8897","3":"1103"},{"1":"S2","2":"8870","3":"1130"},{"1":"S3","2":"2733","3":"267"},{"1":"S4","2":"1703","3":"97"},{"1":"S5","2":"2065","3":"135"},{"1":"S6","2":"2377","3":"123"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Let’s also add the doublet scores to our coldata tibble.</p>
<pre class="r"><code>doublet_scores_scdbl_tbl &lt;- doublet_scores_scdbl %&gt;%
  purrr::imap_dfr(function(smp, name) {
    tibble::tibble(sample = name,
                   barcode = names(smp[[&quot;score&quot;]]),
                   dbl_score_scdbl = smp[[&quot;score&quot;]],
                   pred_dbl_scdbl = smp[[&quot;classifcation&quot;]])
  })

coldata &lt;- coldata %&gt;%
  left_join(doublet_scores_scdbl_tbl, by=c(&quot;sample&quot;, &quot;barcode&quot;))

coldata %&gt;%
  glimpse()</code></pre>
<pre><code>## Rows: 68,072
## Columns: 7
## $ sample          &lt;chr&gt; &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;…
## $ barcode         &lt;chr&gt; &quot;AAACCTGAGACACTAA-1&quot;, &quot;AAACCTGAGGAGTACC-1&quot;, &quot;AA…
## $ joint_barcode   &lt;chr&gt; &quot;HC1_AAACCTGAGACACTAA-1&quot;, &quot;HC1_AAACCTGAGGAGTACC…
## $ sum_umis        &lt;dbl&gt; 5219, 5048, 5519, 5619, 7326, 644, 4597, 4254, …
## $ detected_genes  &lt;int&gt; 1626, 1366, 1719, 1628, 1698, 382, 1299, 1301, …
## $ dbl_score_scdbl &lt;dbl&gt; 0.0127065051, 0.0638903156, 0.0087266751, 0.010…
## $ pred_dbl_scdbl  &lt;fct&gt; singlet, singlet, singlet, singlet, singlet, si…</code></pre>
</div>
<div id="qc-statistics" class="section level1">
<h1>QC Statistics</h1>
<p>Let’s look at some common QC statistics:</p>
<ol style="list-style-type: decimal">
<li>Total number of UMIs per cell</li>
<li>Total number of detected genes per cell</li>
<li>Fraction of mitochondrial genes.</li>
</ol>
<pre class="r"><code>mito_genes &lt;- 
  rownames(rna_list[[1]])[grep(&quot;^MT-&quot;, rownames(rna_list[[1]]))]
mito_genes</code></pre>
<pre><code>##  [1] &quot;MT-ND1&quot;  &quot;MT-ND2&quot;  &quot;MT-CO1&quot;  &quot;MT-CO2&quot;  &quot;MT-ATP8&quot; &quot;MT-ATP6&quot;
##  [7] &quot;MT-CO3&quot;  &quot;MT-ND3&quot;  &quot;MT-ND4L&quot; &quot;MT-ND4&quot;  &quot;MT-ND5&quot;  &quot;MT-ND6&quot; 
## [13] &quot;MT-CYB&quot;</code></pre>
<p>Computing the QC stats.</p>
<p>Looking at the bottom/left tail of the distributions one might argue that the cutoff with regards to the number of UMIs is a little bit too strict for sample “sm023.”</p>
<pre class="r"><code>qc_stats_before &lt;- purrr::imap_dfr(rna_list, function(mtx, name) {
  umis_per_cell &lt;-  Matrix::colSums(mtx)
  genes_per_cell &lt;-  Matrix::colSums(mtx != 0)
  mt_fraction &lt;-  Matrix::colSums(mtx[mito_genes, ]) /  Matrix::colSums(mtx) * 100
  tibble(sample = name, barcode = colnames(mtx), umis_per_cell = umis_per_cell,
         genes_per_cell = genes_per_cell, mt_fraction = mt_fraction)
})

qc_stats_before %&gt;%
  pivot_longer(cols = !c(sample, barcode), names_to = &quot;stat&quot;) %&gt;%
  mutate(stat = ifelse(stat==&quot;mt_fraction&quot;, &quot;mt_fraction (%)&quot;, stat)) %&gt;%
  ggplot() +
  geom_violin(aes(x=sample, y=value, fill=sample)) +
  facet_wrap(~ stat, scales = &quot;free&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = &quot;none&quot;)</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-20-1.png" width="960" /></p>
<p>We will be used the same thresholds as the authors of the paper.</p>
<p>Removing cells with:</p>
<ol style="list-style-type: decimal">
<li>Less than 200 detected genes</li>
<li>More than 6000 detected genes</li>
<li>Fewer than 1000 UMIs</li>
<li>More than 10% mitochondrial fraction</li>
</ol>
<p>Let’s add the QC thresholds to our coldata.</p>
<pre class="r"><code>coldata &lt;- coldata %&gt;%
  left_join(qc_stats_before, by=c(&quot;sample&quot;, &quot;barcode&quot;)) %&gt;%
  mutate(sum_umis_in_bound = (1000 &lt; umis_per_cell),
         detected_genes_in_bound = (20 &lt; genes_per_cell &amp; genes_per_cell &lt;= 6000),
         mt_fraction_in_bound = (mt_fraction &lt; 10)) %&gt;%
  rowwise() %&gt;%
  mutate(qc_pass = all(sum_umis_in_bound, detected_genes_in_bound,
                       mt_fraction_in_bound)) %&gt;%
  ungroup()

coldata %&gt;%
  dplyr::filter(qc_pass) %&gt;%
  glimpse()</code></pre>
<pre><code>## Rows: 51,440
## Columns: 14
## $ sample                  &lt;chr&gt; &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1&quot;, &quot;HC1…
## $ barcode                 &lt;chr&gt; &quot;AAACCTGAGACACTAA-1&quot;, &quot;AAACCTGAGGAGTACC…
## $ joint_barcode           &lt;chr&gt; &quot;HC1_AAACCTGAGACACTAA-1&quot;, &quot;HC1_AAACCTGA…
## $ sum_umis                &lt;dbl&gt; 5219, 5048, 5519, 5619, 7326, 4597, 425…
## $ detected_genes          &lt;int&gt; 1626, 1366, 1719, 1628, 1698, 1299, 130…
## $ dbl_score_scdbl         &lt;dbl&gt; 0.012706505, 0.063890316, 0.008726675, …
## $ pred_dbl_scdbl          &lt;fct&gt; singlet, singlet, singlet, singlet, sin…
## $ umis_per_cell           &lt;dbl&gt; 5219, 5048, 5519, 5619, 7326, 4597, 425…
## $ genes_per_cell          &lt;int&gt; 1626, 1366, 1719, 1628, 1698, 1299, 130…
## $ mt_fraction             &lt;dbl&gt; 1.552021, 1.802694, 1.630730, 1.957644,…
## $ sum_umis_in_bound       &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
## $ detected_genes_in_bound &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
## $ mt_fraction_in_bound    &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
## $ qc_pass                 &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…</code></pre>
<p>And subset the count matrix according to the thresholds</p>
<pre class="r"><code>rna_list_filtered &lt;- purrr::map(names(rna_list), function(smp) {
  mtx &lt;- rna_list[[smp]]
  valid_barcodes &lt;- coldata %&gt;% 
    dplyr::filter(sample == smp, qc_pass) %&gt;% 
    dplyr::pull(barcode)
  cat(paste0(&quot;Removing &quot;, ncol(mtx)-length(valid_barcodes), &quot; cells from &quot;,
             ncol(mtx), &quot; cells in sample &quot;, smp, &quot;\n\n&quot;))
  mtx[,valid_barcodes]
}) %&gt;% set_names(names(rna_list))</code></pre>
<pre><code>## Removing 1537 cells from 10000 cells in sample HC1
## 
## Removing 1816 cells from 10000 cells in sample HC2
## 
## Removing 6420 cells from 8972 cells in sample HC3
## 
## Removing 174 cells from 1800 cells in sample HC4
## 
## Removing 229 cells from 3000 cells in sample M1
## 
## Removing 149 cells from 1800 cells in sample M2
## 
## Removing 2657 cells from 3000 cells in sample M3
## 
## Removing 542 cells from 10000 cells in sample S1
## 
## Removing 229 cells from 10000 cells in sample S2
## 
## Removing 1735 cells from 3000 cells in sample S3
## 
## Removing 320 cells from 1800 cells in sample S4
## 
## Removing 192 cells from 2200 cells in sample S5
## 
## Removing 632 cells from 2500 cells in sample S6</code></pre>
<pre class="r"><code>stopifnot(sum(purrr::map_dbl(rna_list_filtered, ~ ncol(.x))) == sum(coldata$qc_pass))</code></pre>
<pre class="r"><code>qc_stats_after &lt;- purrr::imap_dfr(rna_list_filtered, function(mtx, name) {
  umis_per_cell &lt;- Matrix::colSums(mtx)
  genes_per_cell &lt;- Matrix::colSums(mtx != 0)
  mt_fraction &lt;- Matrix::colSums(mtx[mito_genes, ]) / Matrix::colSums(mtx) * 100
  tibble(sample = name, barcode = colnames(mtx), umis_per_cell = umis_per_cell,
         genes_per_cell = genes_per_cell, mt_fraction = mt_fraction)
})

qc_stats_after %&gt;%
  pivot_longer(cols = !c(sample, barcode), names_to = &quot;stat&quot;) %&gt;%
  mutate(stat = ifelse(stat==&quot;mt_fraction&quot;, &quot;mt_fraction (%)&quot;, stat)) %&gt;%
  ggplot() +
  geom_violin(aes(x=sample, y=value, fill=sample)) +
  facet_wrap(~ stat, scales = &quot;free&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = &quot;none&quot;)</code></pre>
<p><img src="cov19_balf_gex_qc_files/figure-html/unnamed-chunk-23-1.png" width="960" /></p>
</div>
<div id="removing-doublets" class="section level1">
<h1>Removing Doublets</h1>
<p>Lastly we will remove putative doublets. Therefore we will mainly use the <code>scDblFinder</code> Doublet Score (since it is very confident with the doublet assignment and performed goo din the latest benchmark).</p>
<pre class="r"><code>coldata &lt;- coldata %&gt;%
  mutate(is_doublet = ifelse(pred_dbl_scdbl==&quot;doublet&quot;, TRUE, FALSE))

rna_list_filtered &lt;- purrr::map(names(rna_list_filtered), function(smp) {
  mtx &lt;- rna_list_filtered[[smp]]
  valid_barcodes &lt;- coldata %&gt;% 
    dplyr::filter(sample == smp, qc_pass, !is_doublet) %&gt;% 
    dplyr::pull(barcode)
  cat(paste0(&quot;Removing &quot;, ncol(mtx)-length(valid_barcodes), &quot; cells from &quot;,
             ncol(mtx), &quot; cells in sample &quot;, smp, &quot;\n\n&quot;))
  mtx[,valid_barcodes]
}) %&gt;% set_names(names(rna_list))</code></pre>
<pre><code>## Removing 1121 cells from 8463 cells in sample HC1
## 
## Removing 864 cells from 8184 cells in sample HC2
## 
## Removing 172 cells from 2552 cells in sample HC3
## 
## Removing 106 cells from 1626 cells in sample HC4
## 
## Removing 141 cells from 2771 cells in sample M1
## 
## Removing 126 cells from 1651 cells in sample M2
## 
## Removing 11 cells from 343 cells in sample M3
## 
## Removing 1064 cells from 9458 cells in sample S1
## 
## Removing 1118 cells from 9771 cells in sample S2
## 
## Removing 143 cells from 1265 cells in sample S3
## 
## Removing 75 cells from 1480 cells in sample S4
## 
## Removing 120 cells from 2008 cells in sample S5
## 
## Removing 90 cells from 1868 cells in sample S6</code></pre>
<p>How many cells are we removing from each sample?</p>
<pre class="r"><code>coldata_filtered &lt;- coldata %&gt;%
  dplyr::filter(qc_pass, !is_doublet)</code></pre>
</div>
<div id="saving" class="section level1">
<h1>Saving</h1>
<pre class="r"><code>saveRDS(rna_list, 
        paste0(output_dir, timestamp, &quot;_rna_list.RDS&quot;))
saveRDS(rna_list_filtered, 
        paste0(output_dir, timestamp, &quot;_rna_list_filtered.RDS&quot;))
saveRDS(coldata, 
        paste0(output_dir, timestamp, &quot;_coldata_raw.RDS&quot;))
saveRDS(coldata_filtered, 
        paste0(output_dir, timestamp, &quot;_coldata_filtered_qc.RDS&quot;))

save.image(file = paste0(output_dir, &quot;Latest_RNA_QC.RData&quot;))</code></pre>
</div>
<div id="references" class="section level1">
<h1>References</h1>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
